modules/configloader/src/BoardConf.cxx

```diff
@@ -0,0 +1,147 @@
+/*
+ *  BoardConf.cxx
+ *  Copyright 2020 ItJustWorksTM
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include <concepts>
+#include <fstream>
+#include <optional>
+#include <nameof.hpp>
+#include <range/v3/algorithm/transform.hpp>
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wambiguous-reversed-operator"
+#include <rapidjson/document.h>
+#include <rapidjson/istreamwrapper.h>
+#include <rapidjson/schema.h>
+#pragma clang diagnostic pop
+
+#include "BoardConf.hxx"
+
+template <class T, class C>
+struct MptrKeyPair {
+    using FieldType = T;
+    using ClassType = C;
+
+    T C::* mptr;
+    const char* key;
+
+    constexpr MptrKeyPair(T C::* mptr, const char* key) noexcept : mptr{mptr}, key{key} {}
+};
+
+template <class T>
+concept MpKPair = std::is_same_v<T, MptrKeyPair<typename T::FieldType, typename  T::ClassType>>;
+
+#define ADDR_AND_NAME(var) MptrKeyPair{&var, NAMEOF(var).c_str()}
+
+
+template <std::integral T>
+T json_deserialize_integral_unchecked(const rapidjson::Value& jval){
+    if constexpr (std::is_same_v<T, bool>)
+        return jval.GetBool();
+    else if constexpr (std::is_same_v<T, std::uint16_t>)
+        return static_cast<T>(jval.GetUint());
+    else if constexpr (std::is_same_v<T, std::uint32_t>)
+        return jval.GetUint();
+    else if constexpr (std::is_same_v<T, std::uint64_t>)
+        return jval.GetUint64();
+    else if constexpr (std::is_same_v<T, std::int16_t>)
+        return static_cast<T>(jval.GetInt());
+    else if constexpr (std::is_same_v<T, std::int32_t>)
+        return jval.GetInt();
+    else if constexpr (std::is_same_v<T, std::int64_t>)
+        return jval.GetInt64();
+    else
+        static_assert(std::is_void_v<std::void_t<T>>, "Unsupported type");
```

Comment:
One might ask "why `if`[...]`else if`[...]`else if` when your only statement is `return` anyway?"
The answer lies in the fact that these are all `if constexpr` and the `else` body is a `static_assert` which will always fire when evaluated; since post-`return` unreachable code is still evaluated, we need to keep the chain of `if constexpr`s.

Someone who digs into rapidjson might even question the usefulness of this function, since the member function template `rapidjson::GenericValue::Get` exists. The problem with that method is that it does type-checks against `int` and `unsigned` instead of fixed-width integral types, which is problematic for two reasons:
- On most platforms, `sizeof(int) > sizeof(std::int16_t)`, which means that a narrowing conversion is required, and therefore a `static_cast` to make it clean.
- We cannot directly pass `T` as a template parameter since we make the difference between 16 and 32 bit integral types.